api/v1beta1/types_template.go

```diff
@@ -18,9 +18,286 @@ package v1beta1
+	// +optional
+	IdentityRef *corev1.ObjectReference `json:"identityRef,omitempty"`
+
+	// AadProfile is Azure Active Directory configuration to integrate with AKS for aad authentication.
+	// +optional
+	AADProfile *AADProfile `json:"aadProfile,omitempty"`
+
+	// AddonProfiles are the profiles of managed cluster add-on.
+	// +optional
+	AddonProfiles []AddonProfile `json:"addonProfiles,omitempty"`
+
+	// SKU is the SKU of the AKS to be provisioned.
+	// +optional
+	SKU *AKSSku `json:"sku,omitempty"`
+
+	// LoadBalancerProfile is the profile of the cluster load balancer.
+	// +optional
+	LoadBalancerProfile *LoadBalancerProfile `json:"loadBalancerProfile,omitempty"`
+
+	// APIServerAccessProfile is the access profile for AKS API server.
+	// +optional
+	APIServerAccessProfile *APIServerAccessProfileTemplate `json:"apiServerAccessProfile,omitempty"`
+
+	// AutoscalerProfile is the parameters to be applied to the cluster-autoscaler when enabled
+	// +optional
+	AutoScalerProfile *AutoScalerProfile `json:"autoscalerProfile,omitempty"`
+}
+
+// AzureManagedControlPlaneTemplateMachineTemplate specifies an Azure managed control plane template.
+type AzureManagedControlPlaneTemplateMachineTemplate struct {
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	ObjectMeta clusterv1.ObjectMeta `json:"metadata,omitempty"`
+
+	// NodeDrainTimeout is the total amount of time that the controller will spend on draining a controlplane node
+	// The default value is 0, meaning that the node can be drained without any time limitations.
+	// NOTE: NodeDrainTimeout is different from `kubectl drain --timeout`
+	// +optional
+	NodeDrainTimeout *metav1.Duration `json:"nodeDrainTimeout,omitempty"`
+
+	// NodeVolumeDetachTimeout is the total amount of time that the controller will spend on waiting for all volumes
+	// to be detached. The default value is 0, meaning that the volumes can be detached without any time limitations.
+	// +optional
+	NodeVolumeDetachTimeout *metav1.Duration `json:"nodeVolumeDetachTimeout,omitempty"`
+
+	// NodeDeletionTimeout defines how long the machine controller will attempt to delete the Node that the Machine
+	// hosts after the Machine is marked for deletion. A duration of 0 will retry deletion indefinitely.
+	// If no value is provided, the default value for this property of the Machine resource will be used.
+	// +optional
+	NodeDeletionTimeout *metav1.Duration `json:"nodeDeletionTimeout,omitempty"`
+}
+
+// AzureManagedMachinePoolTemplateResourceSpec specifies an Azure managed control plane template resource.
+type AzureManagedMachinePoolTemplateResourceSpec struct {
+	// AdditionalTags is an optional set of tags to add to Azure resources managed by the
+	// Azure provider, in addition to the ones added by default.
+	// +optional
+	AdditionalTags Tags `json:"additionalTags,omitempty"`
```

Comment:
Users can modify any field including AdditionalTags [via patches](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20210526-cluster-class-and-managed-topologies.md#create-a-new-clusterclass-with-patches), although I'm not sure if AdditionalTags is the best type to include in the template if it is something that is unique per resource.

Reply:
Are the fields to be included in template types only things that \*definitely are not\* unique to each cluster or can they include things like this that \*could possibly be\* unique to each cluster if the user sets things up that way? If I want to tag all my clusters with `foo: bar`, it seems like I should be able to do that with a template.

Reply:
I think the convention is that we only exclude types that would have to be unique to each cluster, rather than things that can be possibly unique. Otherwise, if a user wants to template a type that doesn't have to be unique and it's not included as a type, it would be much more difficult than patching a type that is included in the template.